File: assets/js/ELTT-wallet.js
// assets/js/ELTT-wallet.js

"use strict";

/*
 * Frontend-Wallet-Logik für die ELTT-Blockchain.
 * Deterministisch, auditierbar, paritätisch, souveränitätsfokussiert.
 * Die Owner-Wallet entsteht ausschließlich im Frontend und ist nicht hartkodiert.
 */

const ELTTWallet = (() => {
    const {
        TokenKind,
        TxKind,
        TokenType,
        Wallet,
        Transaction,
        Blockchain,
        initBlockchain,
        createGenesis,
        appendBlock,
        computeTxEnergy,
        createWallet,
        getWallet,
        listWalletTokens,
        buildTransferTx,
    } = window.ELTTBlockchain;

    let blockchain = null;
    let currentWalletAddress = null;

    function ensureBlockchain() {
        if (!blockchain) {
            blockchain = initBlockchain();
            createGenesis(blockchain);
        }
        return blockchain;
    }

    function setCurrentWalletAddress(address) {
        currentWalletAddress = address;
        const bc = ensureBlockchain();
        createWallet(bc, address);
    }

    function getCurrentWalletAddress() {
        return currentWalletAddress;
    }

    function getCurrentWallet() {
        if (!currentWalletAddress) {
            return null;
        }
        const bc = ensureBlockchain();
        return getWallet(bc, currentWalletAddress);
    }

    function getWalletState(address) {
        const bc = ensureBlockchain();
        const w = getWallet(bc, address);
        if (!w) {
            return null;
        }
        const tokens = listWalletTokens(w).map(([t, bal]) => ({
            symbol: t.symbol,
            name: t.name,
            kind: t.kind,
            balance: bal,
        }));
        return {
            address: w.address,
            tokens,
        };
    }

    function buildDeterministicTransfer(toAddress, amount, tokenIndex, memo) {
        if (!currentWalletAddress) {
            throw new Error("No current wallet address set.");
        }
        const tx = buildTransferTx(
            currentWalletAddress,
            toAddress,
            amount,
            tokenIndex,
            memo || ""
        );
        computeTxEnergy(tx);
        return tx;
    }

    function applyTransferLocally(tx) {
        const bc = ensureBlockchain();
        const from = getWallet(bc, tx.fromAddress);
        const to = getWallet(bc, tx.toAddress);

        if (!from || !to) {
            return;
        }
        if (tx.tokenIndex < 0 || tx.tokenIndex >= from.balances.length) {
            return;
        }

        const idx = tx.tokenIndex;
        const amount = tx.amount;

        if (from.balances[idx] < amount) {
            return;
        }

        from.balances[idx] -= amount;
        to.balances[idx] += amount;
    }

    function commitTransactions(transactions, timestamp) {
        const bc = ensureBlockchain();
        const txs = transactions.slice();
        for (const tx of txs) {
            if (!tx.energy || tx.energy === 0.0) {
                computeTxEnergy(tx);
            }
        }
        const block = appendBlock(bc, txs, timestamp);
        return block;
    }

    function exportState() {
        const bc = ensureBlockchain();
        return JSON.parse(JSON.stringify(bc));
    }

    return Object.freeze({
        TokenKind,
        TxKind,
        TokenType,
        Wallet,
        Transaction,
        Blockchain,
        ensureBlockchain,
        setCurrentWalletAddress,
        getCurrentWalletAddress,
        getCurrentWallet,
        getWalletState,
        buildDeterministicTransfer,
        applyTransferLocally,
        commitTransactions,
        exportState,
    });
})();

window.ELTTWallet = ELTTWallet;

File: PYTHON-ELTT-MODULE-PARITY/ELTT-Blockchain/ELTT-wallet.py
# PYTHON-ELTT-MODULE-PARITY/ELTT-Blockchain/ELTT-wallet.py
#
# Paritätsmodul für die Wallet-Sicht der ELTT-Blockchain.
# Diese Datei stellt die vollständige Wallet-Logik als
# Perspektive auf die zentrale Blockchain-Implementierung dar.
#
# Keine Beispiele, keine Interpretationen, keine Abweichungen.
# Logik ist identisch zur zentralen ELTT-Blockchain.py.

from ELTT-Blockchain import (  # type: ignore
    Blockchain,
    Wallet,
    TokenType,
    TokenKind,
    Transaction,
    TxKind,
    init_blockchain,
    create_genesis,
    append_block,
    compute_tx_energy,
)


def create_wallet(bc: Blockchain, address: str) -> Wallet:
    """
    Deterministische Wallet-Erstellung.
    Die Ableitung der Adresse aus der 12-Wort-Phrase erfolgt ausschließlich im Frontend.
    Diese Funktion stellt nur sicher, dass eine Wallet-Struktur im Blockchain-Zustand existiert.
    """
    for w in bc.wallets:
        if w.address == address:
            return w
    w = Wallet(address=address)
    w.token_count = len(bc.token_types)
    w.tokens = list(bc.token_types)
    w.balances = [0.0 for _ in range(len(bc.token_types))]
    bc.wallets.append(w)
    return w


def get_wallet(bc: Blockchain, address: str) -> Wallet | None:
    """
    Deterministischer Zugriff auf eine Wallet.
    """
    for w in bc.wallets:
        if w.address == address:
            return w
    return None


def list_wallet_tokens(wallet: Wallet) -> list[tuple[TokenType, float]]:
    """
    Deterministische Auflistung aller Token-Bestände einer Wallet.
    """
    result: list[tuple[TokenType, float]] = []
    for t, bal in zip(wallet.tokens, wallet.balances):
        result.append((t, bal))
    return result


def build_transfer_tx(from_address: str,
                      to_address: str,
                      amount: float,
                      token_index: int,
                      memo: str = "") -> Transaction:
    """
    Deterministische Erstellung einer Transfer-Transaktion.
    """
    return Transaction(
        from_address=from_address,
        to_address=to_address,
        amount=amount,
        token_index=token_index,
        kind=TxKind.TRANSFER,
        memo=memo[:128],
    )


__all__ = [
    "Blockchain",
    "Wallet",
    "TokenType",
    "TokenKind",
    "Transaction",
    "TxKind",
    "init_blockchain",
    "create_genesis",
    "append_block",
    "compute_tx_energy",
    "create_wallet",
    "get_wallet",
    "list_wallet_tokens",
    "build_transfer_tx",
]
